### PR TITLE
PD-262: Add Glyph Prop to Button Component

### PR DESCRIPTION
**Pull request by dominoweir**
## ✍️ Proposed changes

🎟 _Jira ticket:_ [PD-262](https://jira.mongodb.org/browse/PD-262)

## 🛠 Types of changes

- [ ] Tooling (updates to workspace config or internal tooling)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## ✅ Checklist

- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [X] I have run `yarn changeset` and documented my changes

### CODE REVIEW

File: packages/button/src/Button.story.tsx [hunk 1]
```diff
@@ -45,6 +46,11 @@ storiesOf('Buttons', module)
       { div: 'div', span: 'span', button: 'button' },
       'div',
     );
+    const selectedGlyph = select(
```
Comment (brookescarlett) [inline packages/button/src/Button.story.tsx]: I know that @DesignerDave might disagree with me on this one, but think it could be worth either breaking this out into its own story or bringing this logic into the `default` tab. This behavior isn't related to whether or not it's a custom component, so I think it might be confusing to group this here.

File: packages/button/src/Button.spec.tsx [hunk 2]
```diff
@@ -138,4 +139,14 @@ describe('packages/Button', () => {
 
     expect(buttonComponent.tagName.toLowerCase()).toBe('div');
   });
+
+  test(`renders a button with an arbitrary element passed in glyph prop`, () => {
```
Comment (brookescarlett) [inline packages/button/src/Button.spec.tsx]: What exactly are we testing here? I don't see an expect statement... Perhaps we could do something like 

`const { getByTestId } = render( <Button glyph={<Icon glyph="Trash" />} data-testid="trash-button">My Trash Button</Button>)` and checking to make sure that the trash-button contains the Icon?

File: packages/button/src/Button.tsx [hunk 3]
```diff
@@ -333,6 +352,13 @@ const Button = React.forwardRef((props: ButtonProps, forwardRef) => {
     'children',
   ]);
 
+  const modifiedGlyph =
```
Comment (brookescarlett) [inline packages/button/src/Button.tsx]: Thinking that maybe when we modify the glyph we should also change the size of the Icon based on the size of the button. What are your thoughts on this @DesignerDave?

File: packages/button/src/Button.spec.tsx (hunk 2 above)
Comment (dominoweir) [reply packages/button/src/Button.spec.tsx]: Dave recommended that I add a test checking if something renders when the glyph prop is set to a value other than null. I don't know that I did it correctly as my jest knowledge is a little rusty. I can definitely use your suggestion instead.

Comment (brookescarlett) [reply packages/button/src/Button.spec.tsx]: Feel free to test for that, just be sure to add an assertion statement :)

Comment (dominoweir) [reply packages/button/src/Button.spec.tsx]: Ok, went ahead and fixed it. Thanks for catching that!

Comment (brookescarlett) [reply packages/button/src/Button.spec.tsx]: 👍

File: packages/button/src/Button.tsx (hunk 3 above)
Comment (DesignerDave) [reply packages/button/src/Button.tsx]: I think let's stick with the mocks on this one. We aren't changing the font size either I believe.

Comment (DesignerDave) [reply packages/button/src/Button.tsx]: When we do the flat button upgrade, we can consider making larger changes.

File: packages/button/src/Button.spec.tsx [hunk 4]
```diff
@@ -138,4 +138,16 @@ describe('packages/Button', () => {
...
+    const buttonComponent = render(
+      <Button glyph={<svg />}>My Trash Button</Button>,
+    ).container.firstChild;
+
+    if (!typeIs.element(buttonComponent)) {
+      throw new Error('No element was rendered');
+    }
+
+    expect(buttonComponent.getAttribute('glyph')).not.toBe(null);
```
Comment (brookescarlett) [inline packages/button/src/Button.spec.tsx]: Right now this test is testing if the buttonComponent as rendered in the HTML contains a glyph attribute, which is currently occurring but is not the behavior that we want. I'd suggest that first you make sure that we're not spreading the glyph prop along with {...rest} on the Button component and make sure that this test fails. Then, add another test (perhaps checking that the buttonComponent.innerHTML contains <svg>, but feel free to get more creative than that) to ensure that you're testing that we are rendering the desired svg.

Comment (dominoweir) [reply packages/button/src/Button.spec.tsx]: Ok, I went ahead and modified the component and the test to account for this. I'm now checking if the component's innerHTML contains "svg" instead.

Comment (brookescarlett) [reply packages/button/src/Button.spec.tsx]: :+1

File: packages/button/src/Button.tsx (hunk 3 above)
Comment (brookescarlett) [reply packages/button/src/Button.tsx]: Kk sounds good to me

File: packages/button/src/Button.story.tsx (hunk 1 above)
Comment (brookescarlett) [reply packages/button/src/Button.story.tsx]: Still think it makes sense to move this out of custom component. Can be in glyph or it's own story up to you

Comment (dominoweir) [reply packages/button/src/Button.story.tsx]: ok, this logic has now been moved to the default tab

Comment (brookescarlett) [reply packages/button/src/Button.story.tsx]: 👍

File: packages/button/src/Button.tsx [hunk 5]
```diff
@@ -348,8 +375,11 @@ const Button = React.forwardRef((props: ButtonProps, forwardRef) => {
           //
           // https://www.w3.org/TR/css-flexbox-1/#painting
           z-index: 1;
+          display: flex;
+          align-items: center;
```
Comment (DesignerDave) [inline packages/button/src/Button.tsx]: Given we have several CSS properties here now, we should move this from being an inline style declaration to a variable.

Additionally, would `display: inline-flex` be a better change from the default of `display: inline`? Or is there a particular reason we need this to have block sizing?

File: packages/button/src/Button.story.tsx [hunk 6]
```diff
@@ -34,6 +35,11 @@ storiesOf('Buttons', module)
           null,
         ) as any
       }
+      glyph={
+        <Icon
+          glyph={select('Glyph', Object.keys(glyphs) as Array<any>, 'Edit')}
```
Comment (DesignerDave) [inline packages/button/src/Button.story.tsx]: We probably should handle the glyph knob the same way in both stories (e.g. as a variable or inline).

I also _think_ that TypeScript should actually infer the type for `Object.keys(glyphs)` as being something like `Array<keyof typeof glyphs>`, so we might not need to typecast it.

File: packages/button/src/Button.tsx [hunk 7]
```diff
@@ -331,8 +350,16 @@ const Button = React.forwardRef((props: ButtonProps, forwardRef) => {
     'size',
     'variant',
     'children',
+    'glyph',
   ]);
 
+  const modifiedGlyph =
+    glyph && children
+      ? React.cloneElement(glyph, {
+          className: cx({ [glyphMargins[size]]: glyph !== null }),
```
Comment (DesignerDave) [inline packages/button/src/Button.tsx]: It looks like `glyph` can be undefined, so we should use `glyph != null` instead to handle that case.

File: packages/button/src/Button.story.tsx (hunk 6 above)
Comment (dominoweir) [reply packages/button/src/Button.story.tsx]: You are correct, it does not need to be typecast. The glyph knob no longer appears in both stories- Brooke said to just have it in the default component. I wrote it inline because that's how the rest of the knobs were written for the default component, but I can definitely change that if you would prefer to see it in a variable.
